File: internal/risk/risk.go
// Package risk answers "where does AI-written code carry the most risk?":
// AI lines that landed in critical paths without any evidence of review. It
// turns a percentage into something a lead can act on — the specific paths and
// commits — using only git history and grain's provenance.
package risk

import (
	"regexp"
	"sort"
	"strings"

	"github.com/FrontTribe/grain/internal/config"
	"github.com/FrontTribe/grain/internal/gitlog"
	"github.com/FrontTribe/grain/internal/outcomes"
	"github.com/FrontTribe/grain/internal/signal"
)

// DefaultCritical is the built-in list of security- and money-sensitive path
// fragments, matched case-insensitively against each path segment. A repo's
// human_owned paths and any `critical = [...]` patterns are added on top.
var DefaultCritical = []string{
	"auth", "authn", "authz", "login", "session", "token", "secret", "secrets",
	"crypto", "password", "payment", "payments", "billing", "checkout", "iam",
	"permission", "permissions", "migration", "migrations", "infra", "terraform",
	"k8s", "helm", "deploy", "workflows", "dockerfile",
}

// Path is one critical path's exposure.
type Path struct {
	Path       string `json:"path"`          // the matched pattern
	AILines    int    `json:"ai_lines"`      // AI-written lines landed here
	Unreviewed int    `json:"ai_unreviewed"` // of those, with no review evidence
	Commits    int    `json:"commits"`       // commits that put AI lines here
}

// Hotspot is a single commit that put unreviewed AI lines into a critical path.
type Hotspot struct {
	SHA     string `json:"sha"`
	Subject string `json:"subject"`
	Path    string `json:"path"`
	AILines int    `json:"ai_lines"`
}

// Summary is the risk block of a report.
type Summary struct {
	AILines            int       `json:"ai_lines"`               // all AI-written lines in range
	AIUnreviewed       int       `json:"ai_unreviewed"`          // anywhere, no review evidence
	CriticalAILines    int       `json:"critical_ai_lines"`      // in critical paths
	CriticalUnreviewed int       `json:"critical_ai_unreviewed"` // the headline number
	Critical           []Path    `json:"critical"`               // per path, worst first
	Top                []Hotspot `json:"top"`                    // worst commits, up to 10
	Patterns           []string  `json:"patterns"`               // what counted as critical
	ReviewedCommits    int       `json:"reviewed_commits"`
	UnreviewedCommits  int       `json:"unreviewed_commits"`
}

// UnreviewedShare is the share of AI lines in critical paths that had no
// review evidence (0..1); 0 when there are none.
func (s Summary) UnreviewedShare() float64 {
	if s.CriticalAILines == 0 {
		return 0
	}
	return float64(s.CriticalUnreviewed) / float64(s.CriticalAILines)
}

var (
	reviewedTrailer = regexp.MustCompile(`(?im)^\s*(reviewed-by|reviewed-on|approved-by):`)
	squashSubject   = regexp.MustCompile(`\(#\d+\)\s*$`)
	mergeSubject    = regexp.MustCompile(`^Merge (pull request|branch)\b`)
)

// Reviewed reports whether a commit carries evidence of having gone through
// review. Git can't see approvals, so this is the union of the signals it does
// leave: a Reviewed-by/Approved-by trailer, a squash-merge subject "(#123)", a
// merge subject, arriving via a merge (not on the first-parent chain), or being
// applied by someone other than the author (committer ≠ author, as with
// GitHub's merge button). Absence is "no evidence", not proof of no review.
func Reviewed(c gitlog.Commit, firstParent map[string]bool) bool {
	if reviewedTrailer.MatchString(c.Body) {
		return true
	}
	subj := strings.TrimSpace(c.Subject)
	if squashSubject.MatchString(subj) || mergeSubject.MatchString(subj) {
		return true
	}
	if len(firstParent) > 0 && !firstParent[c.SHA] {
		return true // came in through a merge
	}
	if c.CommitterEmail != "" && c.AuthorEmail != "" && !strings.EqualFold(c.CommitterEmail, c.AuthorEmail) {
		return true
	}
	return false
}

// Matches reports which pattern (if any) a file path is critical under.
// Explicit patterns (human_owned, critical) match as path prefixes the way the
// policy does ("src/auth/**"); built-in fragments match whole path segments or
// file stems, case-insensitively, so "auth" hits src/auth/ and auth.go but not
// author.go.
func Matches(path string, patterns []string) (string, bool) {
	lower := strings.ToLower(path)
	segs := strings.Split(lower, "/")
	for _, pat := range patterns {
		p := strings.ToLower(strings.TrimSpace(pat))
		if p == "" {
			continue
		}
		if strings.ContainsAny(p, "/*") {
			base := strings.TrimSuffix(strings.TrimSuffix(strings.TrimSuffix(p, "/**"), "/*"), "/")
			if base != "" && (lower == base || strings.HasPrefix(lower, base+"/")) {
				return pat, true
			}
			continue
		}
		for _, seg := range segs {
			stem := seg
			if i := strings.LastIndexByte(seg, '.'); i > 0 {
				stem = seg[:i]
			}
			if seg == p || stem == p {
				return pat, true
			}
		}
	}
	return "", false
}

// Compute walks the commits and attributes every AI-written line to the file
// it landed in, then rolls up by critical pattern. `added` is the per-SHA diff
// from gitlog (needed to tell which lines are AI when attestation is line-level).
func Compute(commits []gitlog.Commit, added map[string]map[string][]string, firstParent map[string]bool, patterns []string, cfg config.Config) Summary {
	if len(patterns) == 0 {
		patterns = DefaultCritical
	}
	sum := Summary{Patterns: patterns}
	byPath := map[string]*Path{}
	seenPathCommit := map[string]bool{}
	var hot []Hotspot

	for _, c := range commits {
		s := signal.Extract(c, cfg)
		hashes := outcomes.AIHashes(c.Note)
		lineLevel := len(hashes) > 0
		wholeAI := !lineLevel && (s.AttestedClass == "ai" || s.DeclaredAI)
		if !lineLevel && !wholeAI {
			continue // no AI lines to attribute
		}
		reviewed := Reviewed(c, firstParent)
		if reviewed {
			sum.ReviewedCommits++
		} else {
			sum.UnreviewedCommits++
		}

		files := added[c.SHA]
		for path, lines := range files {
			n := 0
			for _, l := range lines {
				if !outcomes.Substantive(l) {
					continue
				}
				if wholeAI || hashes[outcomes.LineHash(l)] {
					n++
				}
			}
			if n == 0 {
				continue
			}
			sum.AILines += n
			if !reviewed {
				sum.AIUnreviewed += n
			}
			pat, crit := Matches(path, patterns)
			if !crit {
				continue
			}
			sum.CriticalAILines += n
			ps := byPath[pat]
			if ps == nil {
				ps = &Path{Path: pat}
				byPath[pat] = ps
			}
			ps.AILines += n
			if !seenPathCommit[pat+"\x00"+c.SHA] {
				seenPathCommit[pat+"\x00"+c.SHA] = true
				ps.Commits++
			}
			if !reviewed {
				sum.CriticalUnreviewed += n
				ps.Unreviewed += n
				hot = append(hot, Hotspot{SHA: c.SHA, Subject: strings.TrimSpace(c.Subject), Path: pat, AILines: n})
			}
		}
	}

	for _, ps := range byPath {
		sum.Critical = append(sum.Critical, *ps)
	}
	sort.Slice(sum.Critical, func(i, j int) bool {
		if sum.Critical[i].Unreviewed != sum.Critical[j].Unreviewed {
			return sum.Critical[i].Unreviewed > sum.Critical[j].Unreviewed
		}
		return sum.Critical[i].AILines > sum.Critical[j].AILines
	})
	sort.Slice(hot, func(i, j int) bool { return hot[i].AILines > hot[j].AILines })
	if len(hot) > 10 {
		hot = hot[:10]
	}
	sum.Top = hot
	return sum
}

File: web/src/lib/risk.ts
// Cloud-side risk analysis: AI-written lines that landed in critical paths
// without review evidence, over the commits a GitHub scan deep-scanned. Mirrors
// internal/risk in the engine, with one upgrade only the Cloud can make — review
// evidence comes from the GitHub PR API (an associated pull request, and whether
// someone other than the author approved it), not just from what git leaves.
import { createHash } from "crypto";

// Same built-in list as internal/risk.DefaultCritical.
export const DEFAULT_CRITICAL = [
  "auth", "authn", "authz", "login", "session", "token", "secret", "secrets",
  "crypto", "password", "payment", "payments", "billing", "checkout", "iam",
  "permission", "permissions", "migration", "migrations", "infra", "terraform",
  "k8s", "helm", "deploy", "workflows", "dockerfile",
];

// Identical to outcomes.LineHash in the engine: sha256 of the trimmed line,
// first 10 hex chars. Attestation notes carry these, so they must match exactly.
export function lineHash(line: string): string {
  return createHash("sha256").update(line.trim()).digest("hex").slice(0, 10);
}

export function substantive(line: string): boolean {
  return line.trim().length > 3;
}

// Parse the AI-Hashes line of a grain note into a set (null when absent).
export function parseAIHashes(note: string): Set<string> | null {
  for (const raw of note.split("\n")) {
    const idx = raw.indexOf(":");
    if (idx < 0) continue;
    if (raw.slice(0, idx).trim().toLowerCase() !== "ai-hashes") continue;
    const set = new Set<string>();
    for (const h of raw.slice(idx + 1).split(",")) {
      const t = h.trim();
      if (t) set.add(t);
    }
    return set.size ? set : null;
  }
  return null;
}

// Which pattern (if any) a path is critical under. Explicit patterns with a
// "/" or "*" match as prefixes ("src/auth/**"); bare fragments match a whole
// path segment or file stem, case-insensitively ("auth" hits src/auth/ and
// auth.go, not author.go). Same rules as the engine.
export function matchesCritical(path: string, patterns: string[]): string | null {
  const lower = path.toLowerCase();
  const segs = lower.split("/");
  for (const pat of patterns) {
    const p = pat.trim().toLowerCase();
    if (!p) continue;
    if (/[/*]/.test(p)) {
      const base = p.replace(/\/\*\*$/, "").replace(/\/\*$/, "").replace(/\/$/, "");
      if (base && (lower === base || lower.startsWith(base + "/"))) return pat;
      continue;
    }
    for (const seg of segs) {
      const dot = seg.lastIndexOf(".");
      const stem = dot > 0 ? seg.slice(0, dot) : seg;
      if (seg === p || stem === p) return pat;
    }
  }
  return null;
}

export type ReviewEvidence = "approved" | "pr" | "heuristic" | "none";

// What git itself leaves behind: a reviewer trailer, a squash "(#123)" or merge
// subject, or a committer other than the author (GitHub's merge button, a
// maintainer applying a patch).
export function heuristicReview(c: { message?: string; authorEmail?: string; committerEmail?: string }): boolean {
  const msg = c.message ?? "";
  if (/^\s*(reviewed-by|reviewed-on|approved-by):/im.test(msg)) return true;
  const subject = msg.split("\n")[0]?.trim() ?? "";
  if (/\(#\d+\)\s*$/.test(subject) || /^Merge (pull request|branch)\b/.test(subject)) return true;
  const a = (c.authorEmail ?? "").toLowerCase();
  const m = (c.committerEmail ?? "").toLowerCase();
  if (a && m && a !== m) return true;
  return false;
}

// Ask GitHub whether a commit belongs to a pull request and whether that PR
// carries an approval from someone other than its author. Best-effort.
export async function prEvidence(
  owner: string,
  repo: string,
  sha: string,
  headers: Record<string, string>,
): Promise<{ pr: number | null; approved: boolean }> {
  const base = `https://api.github.com/repos/${encodeURIComponent(owner)}/${encodeURIComponent(repo)}`;
  try {
    const res = await fetch(`${base}/commits/${encodeURIComponent(sha)}/pulls?per_page=5`, { headers, cache: "no-store" });
    if (!res.ok) return { pr: null, approved: false };
    const prs = (await res.json()) as { number?: number; user?: { login?: string } }[];
    if (!Array.isArray(prs) || prs.length === 0) return { pr: null, approved: false };
    const first = prs[0];
    const number = first.number ?? null;
    if (number == null) return { pr: null, approved: false };
    const rv = await fetch(`${base}/pulls/${number}/reviews?per_page=50`, { headers, cache: "no-store" });
    if (!rv.ok) return { pr: number, approved: false };
    const reviews = (await rv.json()) as { state?: string; user?: { login?: string } }[];
    const author = first.user?.login;
    const approved = Array.isArray(reviews) && reviews.some((r) => r.state === "APPROVED" && r.user?.login && r.user.login !== author);
    return { pr: number, approved };
  } catch {
    return { pr: null, approved: false };
  }
}

export type CloudRiskCommit = {
  sha: string;
  message: string;
  authorEmail?: string;
  committerEmail?: string;
  added: Record<string, string[]> | null; // per-file added lines
  aiHashes: Set<string> | null; // line-level attestation, when present
  wholeAI: boolean; // attested/declared as a whole
};

export type CloudRiskPath = { path: string; ai_lines: number; ai_unreviewed: number; commits: number };
export type CloudRiskHotspot = { sha: string; subject: string; path: string; ai_lines: number; evidence: ReviewEvidence };

export type CloudRisk = {
  ai_lines: number;
  ai_unreviewed: number;
  critical_ai_lines: number;
  critical_ai_unreviewed: number;
  approved_ai_lines: number; // critical AI lines whose PR carries an approval
  critical: CloudRiskPath[];
  top: CloudRiskHotspot[];
  patterns: string[];
  reviewed_commits: number;
  unreviewed_commits: number;
  commits: number; // the window examined
  source: "cloud";
};

async function mapLimit<T, R>(items: T[], limit: number, fn: (t: T) => Promise<R>): Promise<R[]> {
  const out: R[] = new Array(items.length);
  let i = 0;
  await Promise.all(
    Array.from({ length: Math.min(limit, items.length) }, async () => {
      while (i < items.length) {
        const idx = i++;
        out[idx] = await fn(items[idx]);
      }
    }),
  );
  return out;
}

// Compute the risk block for a deep-scan window. PR lookups are made only for
// commits that actually put AI lines into a critical path and have no
// heuristic evidence — the only ones whose answer changes the headline.
export async function computeCloudRisk(opts: {
  owner: string;
  repo: string;
  headers: Record<string, string>;
  commits: CloudRiskCommit[];
  patterns?: string[];
}): Promise<{ risk: CloudRisk; evidence: Map<string, ReviewEvidence> }> {
  const patterns = opts.patterns?.length ? opts.patterns : DEFAULT_CRITICAL;
  const risk: CloudRisk = {
    ai_lines: 0, ai_unreviewed: 0, critical_ai_lines: 0, critical_ai_unreviewed: 0, approved_ai_lines: 0,
    critical: [], top: [], patterns, reviewed_commits: 0, unreviewed_commits: 0,
    commits: opts.commits.length, source: "cloud",
  };

  // Pass 1: attribute AI lines per file, per commit.
  type Attr = { c: CloudRiskCommit; total: number; critical: { path: string; pattern: string; n: number }[] };
  const attrs: Attr[] = [];
  for (const c of opts.commits) {
    if (!c.added || (!c.wholeAI && !c.aiHashes)) continue;
    let total = 0;
    const critical: Attr["critical"] = [];
    for (const [path, lines] of Object.entries(c.added)) {
      let n = 0;
      for (const l of lines) {
        if (!substantive(l)) continue;
        if (c.wholeAI || c.aiHashes?.has(lineHash(l))) n++;
      }
      if (n === 0) continue;
      total += n;
      const pattern = matchesCritical(path, patterns);
      if (pattern) critical.push({ path, pattern, n });
    }
    if (total > 0) attrs.push({ c, total, critical });
  }

  // Pass 2: review evidence. Heuristics are free; ask GitHub only where it matters.
  const evidence = new Map<string, ReviewEvidence>();
  const needPR: Attr[] = [];
  for (const a of attrs) {
    if (heuristicReview({ message: a.c.message, authorEmail: a.c.authorEmail, committerEmail: a.c.committerEmail })) {
      evidence.set(a.c.sha, "heuristic");
    } else if (a.critical.length > 0) {
      needPR.push(a);
    } else {
      evidence.set(a.c.sha, "none");
    }
  }
  const prResults = await mapLimit(needPR, 4, (a) => prEvidence(opts.owner, opts.repo, a.c.sha, opts.headers));
  needPR.forEach((a, i) => {
    const r = prResults[i];
    evidence.set(a.c.sha, r.approved ? "approved" : r.pr != null ? "pr" : "none");
  });

  // Pass 3: tally.
  const byPattern = new Map<string, CloudRiskPath>();
  const seen = new Set<string>();
  for (const a of attrs) {
    const ev = evidence.get(a.c.sha) ?? "none";
    const reviewed = ev !== "none";
    if (reviewed) risk.reviewed_commits++;
    else risk.unreviewed_commits++;
    risk.ai_lines += a.total;
    if (!reviewed) risk.ai_unreviewed += a.total;
    for (const { pattern, n } of a.critical) {
      risk.critical_ai_lines += n;
      if (ev === "approved") risk.approved_ai_lines += n;
      const ps = byPattern.get(pattern) ?? { path: pattern, ai_lines: 0, ai_unreviewed: 0, commits: 0 };
      ps.ai_lines += n;
      const key = pattern + "\0" + a.c.sha;
      if (!seen.has(key)) {
        seen.add(key);
        ps.commits++;
      }
      if (!reviewed) {
        risk.critical_ai_unreviewed += n;
        ps.ai_unreviewed += n;
        risk.top.push({ sha: a.c.sha, subject: (a.c.message.split("\n")[0] ?? "").trim(), path: pattern, ai_lines: n, evidence: ev });
      }
      byPattern.set(pattern, ps);
    }
  }
  risk.critical = [...byPattern.values()].sort((x, y) => y.ai_unreviewed - x.ai_unreviewed || y.ai_lines - x.ai_lines);
  risk.top.sort((x, y) => y.ai_lines - x.ai_lines);
  risk.top = risk.top.slice(0, 10);
  // The evidence map is returned so the security pass can reuse it.
  return { risk, evidence };
}
